FeedService/Infrastructure/StorageAccountRepository.cs

```diff
@@ -0,0 +1,126 @@
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Compression;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace FeedService.Infrastructure
+{
+    public class StorageAccountRepository : IStorageAccountRepository
+    {
+        private const string _piletContainer = "pilets";
+
+        public StorageAccountRepository()
+        {
+        }
+
+        public async Task<List<string>> GetLatestPiletsInfoAsync()
+        {
+            var piletsGroupedWithAllVersions = new Dictionary<string, List<string>>();
+            var latestPiletsInfo = new List<string>();
+            BlobServiceClient blobServiceClient = new BlobServiceClient(GetConnectionString());
+            BlobContainerClient containerClient = blobServiceClient.GetBlobContainerClient(_piletContainer);
+            var blobs = containerClient.GetBlobs();
+            await foreach (BlobItem item in containerClient.GetBlobsAsync())
+            {
+                var piletName = GetPiletName(item.Name);
+                if (piletsGroupedWithAllVersions.ContainsKey(piletName))
+                {
+                    piletsGroupedWithAllVersions[piletName].Add(item.Name);
+                }
+                else
+                {
+                    piletsGroupedWithAllVersions.Add(piletName, new List<string> { item.Name });
+                }
+            }
+
+            foreach (var piletGroupedWithAllVersions in piletsGroupedWithAllVersions)
+            {
+                latestPiletsInfo.Add(piletGroupedWithAllVersions.Value.OrderBy(i => i).LastOrDefault());
+            }
+
+
+            return latestPiletsInfo.Distinct().ToList();
+        }
+
+        public async Task UploadPilet(List<PackageFile> packageFiles, PackageJson packageJson, byte[] zippedFiles)
+        {
+            try
+            {
+                BlobContainerClient containerClient = new BlobContainerClient(GetConnectionString(), _piletContainer);
+                BlobClient blobClient = containerClient.GetBlobClient(packageJson.Name + "/" + packageJson.Version + ".zip");
+                using (var st = new MemoryStream(zippedFiles))
+                {
+                    await blobClient.UploadAsync(st, true);
+                }
+            }
+            catch (Exception)
+            {
+
+                throw;
+            }
+        }
+
+        public async Task<string> GetPiletFile(string fileName, string version, string piletName)
+        {
+            var file = "";
+
+            try
+            {
+                BlobServiceClient blobServiceClient = new BlobServiceClient(GetConnectionString());
+                BlobContainerClient containerClient = blobServiceClient.GetBlobContainerClient(_piletContainer);
+                BlobClient bc = containerClient.GetBlobClient(piletName + "/" + version + ".zip");
+                if (await bc.ExistsAsync())
+                {
+                    using (var mem = new MemoryStream())
```

Comment:
Always getting the full ZIP for each file that is demanded sounds overly complex and slow. Any reason for not uploading *all* files from the archive and just getting the demanded file?

Reply:
Right now i am uploading the pilet files as ZIP filder. Because it dosn't make sense to upload single files to storage account as a blob. What i am thinking here to use some kind of intermediate where we keep the files, like Redis cache. Or maybe azure CDN but i am not sure about that yet.

Reply:
> Because it dosn't make sense to upload single files to storage account as a blob.

Can you elaborate why? As far as I know you are not paying by the number of files. But you are paying per compute time on Azure Functions, which increases by having more bytes to download + process (unzip) in this case. Maybe I am wrong here, I'd just love to understand the reasoning.

Reply:
I initially thought that it makes sense to put all files in a zip folder, to save space. But you are right that the computation will take longer And it doesn't make sense that each file downloads the whole zip. Maybe i should upload the file as is in a folder. and then retrieve them one by one. So retrieve only that which is needed.

Reply:
What do you say?

Reply:
I'd say upload them already in the right folder structure (in a folder `{pilet-name}/{pilet-version}`). Then you can retrieve only the demanded file via `{pilet-name}/{pilet-version}/{file-path}`.

Regarding the all pilets problem: Well, even if somebody wants to get all pilets all file requests would (I think) still be individual. So this should not matter much. Or am I missing something?

Reply:
What i meant was the Get call that this returns https://feed.piral.cloud/api/v1/pilet
It will be hard to return the same format when you call individual files.

Reply:
Hm not sure I follow. Calling `https://feed.piral.cloud/api/v1/pilet` returns the metadata where only a single file is linked, which is the main JS file. So this would lead to `{website-host}/api/{pilet-name}/{pilet-version}/{file-path}`, which we can resolve against the blob storage.